pallets/rc-migrator/src/conviction_voting.rs

```diff
@@ -0,0 +1,252 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::*;
+use frame_support::traits::{Currency, Polling};
+use pallet_conviction_voting::{ClassLocksFor, TallyOf, Voting};
+use sp_runtime::traits::Zero;
+
+/// Stage of the scheduler pallet migration.
+#[derive(Encode, Decode, Clone, RuntimeDebug, TypeInfo, MaxEncodedLen, PartialEq, Eq)]
+pub enum ConvictionVotingStage<AccountId, Class> {
+	VotingFor(Option<(AccountId, Class)>),
+	ClassLocksFor(Option<AccountId>),
+	Finished,
+}
+
+#[derive(Encode, Decode, RuntimeDebug, Clone, TypeInfo, MaxEncodedLen, PartialEq, Eq)]
+pub enum RcConvictionVotingMessage<AccountId, Class, Voting, Balance> {
+	VotingFor(AccountId, Class, Voting),
+	ClassLocksFor(AccountId, Vec<(Class, Balance)>),
+}
+
+pub type RcConvictionVotingMessageOf<T> = RcConvictionVotingMessage<
+	<T as frame_system::Config>::AccountId,
+	alias::ClassOf<T>,
+	alias::VotingOf<T>,
+	alias::BalanceOf<T>,
+>;
+
+pub struct ConvictionVotingMigrator<T: Config> {
+	_phantom: sp_std::marker::PhantomData<T>,
+}
+
+impl<T: Config> PalletMigration for ConvictionVotingMigrator<T> {
+	type Key = ConvictionVotingStage<T::AccountId, alias::ClassOf<T>>;
+	type Error = Error<T>;
+
+	fn migrate_many(
+		last_key: Option<Self::Key>,
+		weight_counter: &mut WeightMeter,
+	) -> Result<Option<Self::Key>, Self::Error> {
+		let mut last_key = last_key.unwrap_or(ConvictionVotingStage::VotingFor(None));
+		let mut messages = Vec::new();
+
+		loop {
+			if weight_counter
+				.try_consume(<T as frame_system::Config>::DbWeight::get().reads_writes(1, 1))
+				.is_err()
+			{
+				if messages.is_empty() {
+					return Err(Error::OutOfWeight);
+				} else {
+					break;
+				}
+			}
+			if messages.len() > 10_000 {
+				log::warn!(target: LOG_TARGET, "Weight allowed very big batch, stopping");
+				break;
+			}
+
+			last_key = match last_key {
+				ConvictionVotingStage::VotingFor(last_voting_key) => {
+					let mut iter = match last_voting_key {
+						None => alias::VotingFor::<T>::iter(),
+						Some((account_id, class)) => alias::VotingFor::<T>::iter_from(
+							alias::VotingFor::<T>::hashed_key_for(account_id, class),
+						),
+					};
+					match iter.next() {
+						Some((account_id, class, voting)) => {
+							alias::VotingFor::<T>::remove(&account_id, &class);
+							if Pallet::<T>::is_empty_conviction_vote(&voting) {
+								// from the Polkadot 17.01.2025 snapshot 20575 records
+								// issue: https://github.com/paritytech/polkadot-sdk/issues/7458
+								log::info!(
```

Comment:
We need some good testing to ensure that this does not break other stuff. Will note it.

Reply:
If the issue I created wont be solved before, I will add tests.